tests/test_edgeql_functions.py

```diff
@@ -996,6 +996,309 @@ async def test_edgeql_functions_array_get_08(self):
             [2],
         )
 
+    async def test_edgeql_functions_array_set_01(self):
+        # Positive indexes
+        await self.assert_query_result(
+            r'''SELECT array_set([1, 2, 3, 4], 0, 9);''',
+            [[9, 2, 3, 4]],
+        )
+
+        await self.assert_query_result(
+            r'''SELECT array_set([1, 2, 3, 4], 1, 9);''',
+            [[1, 9, 3, 4]],
+        )
+
+        await self.assert_query_result(
+            r'''SELECT array_set([1, 2, 3, 4], 2, 9);''',
+            [[1, 2, 9, 4]],
+        )
+
+        await self.assert_query_result(
+            r'''SELECT array_set([1, 2, 3, 4], 3, 9);''',
+            [[1, 2, 3, 9]],
+        )
+
+        await self.assert_query_result(
+            r'''SELECT array_set([1, 2, 3, 4], 4, 9);''',
+            [[1, 2, 3, 4]],
+        )
+
+        # Negative indexes
+        await self.assert_query_result(
+            r'''SELECT array_set([1, 2, 3, 4], -1, 9);''',
+            [[1, 2, 3, 9]],
+        )
+
+        await self.assert_query_result(
+            r'''SELECT array_set([1, 2, 3, 4], -2, 9);''',
+            [[1, 2, 9, 4]],
+        )
+
+        await self.assert_query_result(
+            r'''SELECT array_set([1, 2, 3, 4], -3, 9);''',
+            [[1, 9, 3, 4]],
+        )
+
+        await self.assert_query_result(
+            r'''SELECT array_set([1, 2, 3, 4], -4, 9);''',
+            [[9, 2, 3, 4]],
+        )
+
+        await self.assert_query_result(
+            r'''SELECT array_set([1, 2, 3, 4], -5, 9);''',
+            [[1, 2, 3, 4]],
+        )
+
+        # Size 1 array
+        await self.assert_query_result(
+            r'''SELECT array_set([1], 0, 9);''',
+            [[9]],
+        )
+
+        await self.assert_query_result(
+            r'''SELECT array_set([1], 1, 9);''',
+            [[1]],
+        )
+
+        await self.assert_query_result(
+            r'''SELECT array_set([1], -1, 9);''',
+            [[9]],
+        )
+
+        await self.assert_query_result(
+            r'''SELECT array_set([1], -2, 9);''',
+            [[1]],
+        )
+
+        # Size 0 array
+        await self.assert_query_result(
+            r'''SELECT array_set(<array<int64>>[], 0, 9);''',
+            [[]],
+        )
+
+        await self.assert_query_result(
+            r'''SELECT array_set(<array<int64>>[], 1, 9);''',
+            [[]],
+        )
```

Comment:
_Assigning_ a value outside of bounds of an array does not reasonably produce an unchanged array as we don't support gaps in indexes. So getting an error is reasonable here (instead of silently ignoring an undefined operation).
I also assume that for `array_set` the _length_ of the array is invariant, so we cannot just add an element at index 0 of an `[]`.

Whatever bounds-checking policy we have for `array_set` should apply to `array_insert`. In fact, the only difference is that `array_set([0, 1, 2], 3, 9)` is illegal (as there is no element there to set), while `array_insert([0, 1, 2], 3, 9) == [0, 1, 2, 9]` since it is possible to insert an element after the last array element and still end up with a contiguous array. Other than that corner case, both `array_set` and `array_insert` are logically trying to create a non-contiguous array when they are putting elements out of bounds. As far as I can see, `array_insert` behaves reasonably w.r.t. acceptable and unacceptable index values/bound checks.